From 1898944000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Daniel Juenger <sleeepyjack@users.noreply.github.com>
Date: Mon, 14 Apr 2026 00:00:00 +0000
Subject: [PATCH] Fix codegen in 128bit atomic CAS (#8403)

Fix wrong inline asm operand indices in all atom.cas.*.b128 variants.
See https://github.com/NVIDIA/cccl/issues/8402
---
 .../cuda/std/__atomic/functions/cuda_ptx_generated.h | 80 ++++++++++----------
 1 file changed, 40 insertions(+), 40 deletions(-)

diff --git a/libcudacxx/include/cuda/std/__atomic/functions/cuda_ptx_generated.h b/libcudacxx/include/cuda/std/__atomic/functions/cuda_ptx_generated.h
index f3e30d53039..479815f4136 100644
--- a/libcudacxx/include/cuda/std/__atomic/functions/cuda_ptx_generated.h
+++ b/libcudacxx/include/cuda/std/__atomic/functions/cuda_ptx_generated.h
@@ -1585,8 +1585,8 @@ static inline _CCCL_DEVICE bool __cuda_atomic_compare_exchange(
     {
       .reg .b128 _d;
       .reg .b128 _v;
-      mov.b128 _d, {%0, %1};
-      mov.b128 _v, {%4, %5};
+      mov.b128 _d, {%3, %4};
+      mov.b128 _v, {%5, %6};
       atom.cas.acquire.cta.b128 _d,[%2],_d,_v;
       mov.b128 {%0, %1}, _d;
     }
@@ -1604,8 +1604,8 @@ static inline _CCCL_DEVICE bool __cuda_atomic_compare_exchange(
     {
       .reg .b128 _d;
       .reg .b128 _v;
-      mov.b128 _d, {%0, %1};
-      mov.b128 _v, {%4, %5};
+      mov.b128 _d, {%3, %4};
+      mov.b128 _v, {%5, %6};
       atom.cas.acquire.cluster.b128 _d,[%2],_d,_v;
       mov.b128 {%0, %1}, _d;
     }
@@ -1623,8 +1623,8 @@ static inline _CCCL_DEVICE bool __cuda_atomic_compare_exchange(
     {
       .reg .b128 _d;
       .reg .b128 _v;
-      mov.b128 _d, {%0, %1};
-      mov.b128 _v, {%4, %5};
+      mov.b128 _d, {%3, %4};
+      mov.b128 _v, {%5, %6};
       atom.cas.acquire.gpu.b128 _d,[%2],_d,_v;
       mov.b128 {%0, %1}, _d;
     }
@@ -1642,8 +1642,8 @@ static inline _CCCL_DEVICE bool __cuda_atomic_compare_exchange(
     {
       .reg .b128 _d;
       .reg .b128 _v;
-      mov.b128 _d, {%0, %1};
-      mov.b128 _v, {%4, %5};
+      mov.b128 _d, {%3, %4};
+      mov.b128 _v, {%5, %6};
       atom.cas.acquire.sys.b128 _d,[%2],_d,_v;
       mov.b128 {%0, %1}, _d;
     }
@@ -1661,8 +1661,8 @@ static inline _CCCL_DEVICE bool __cuda_atomic_compare_exchange(
     {
       .reg .b128 _d;
       .reg .b128 _v;
-      mov.b128 _d, {%0, %1};
-      mov.b128 _v, {%4, %5};
+      mov.b128 _d, {%3, %4};
+      mov.b128 _v, {%5, %6};
       atom.cas.relaxed.cta.b128 _d,[%2],_d,_v;
       mov.b128 {%0, %1}, _d;
     }
@@ -1680,8 +1680,8 @@ static inline _CCCL_DEVICE bool __cuda_atomic_compare_exchange(
     {
       .reg .b128 _d;
       .reg .b128 _v;
-      mov.b128 _d, {%0, %1};
-      mov.b128 _v, {%4, %5};
+      mov.b128 _d, {%3, %4};
+      mov.b128 _v, {%5, %6};
       atom.cas.relaxed.cluster.b128 _d,[%2],_d,_v;
       mov.b128 {%0, %1}, _d;
     }
@@ -1699,8 +1699,8 @@ static inline _CCCL_DEVICE bool __cuda_atomic_compare_exchange(
     {
       .reg .b128 _d;
       .reg .b128 _v;
-      mov.b128 _d, {%0, %1};
-      mov.b128 _v, {%4, %5};
+      mov.b128 _d, {%3, %4};
+      mov.b128 _v, {%5, %6};
       atom.cas.relaxed.gpu.b128 _d,[%2],_d,_v;
       mov.b128 {%0, %1}, _d;
     }
@@ -1718,8 +1718,8 @@ static inline _CCCL_DEVICE bool __cuda_atomic_compare_exchange(
     {
       .reg .b128 _d;
       .reg .b128 _v;
-      mov.b128 _d, {%0, %1};
-      mov.b128 _v, {%4, %5};
+      mov.b128 _d, {%3, %4};
+      mov.b128 _v, {%5, %6};
       atom.cas.relaxed.sys.b128 _d,[%2],_d,_v;
       mov.b128 {%0, %1}, _d;
     }
@@ -1737,8 +1737,8 @@ static inline _CCCL_DEVICE bool __cuda_atomic_compare_exchange(
     {
       .reg .b128 _d;
       .reg .b128 _v;
-      mov.b128 _d, {%0, %1};
-      mov.b128 _v, {%4, %5};
+      mov.b128 _d, {%3, %4};
+      mov.b128 _v, {%5, %6};
       atom.cas.release.cta.b128 _d,[%2],_d,_v;
       mov.b128 {%0, %1}, _d;
     }
@@ -1756,8 +1756,8 @@ static inline _CCCL_DEVICE bool __cuda_atomic_compare_exchange(
     {
       .reg .b128 _d;
       .reg .b128 _v;
-      mov.b128 _d, {%0, %1};
-      mov.b128 _v, {%4, %5};
+      mov.b128 _d, {%3, %4};
+      mov.b128 _v, {%5, %6};
       atom.cas.release.cluster.b128 _d,[%2],_d,_v;
       mov.b128 {%0, %1}, _d;
     }
@@ -1775,8 +1775,8 @@ static inline _CCCL_DEVICE bool __cuda_atomic_compare_exchange(
     {
       .reg .b128 _d;
       .reg .b128 _v;
-      mov.b128 _d, {%0, %1};
-      mov.b128 _v, {%4, %5};
+      mov.b128 _d, {%3, %4};
+      mov.b128 _v, {%5, %6};
       atom.cas.release.gpu.b128 _d,[%2],_d,_v;
       mov.b128 {%0, %1}, _d;
     }
@@ -1794,8 +1794,8 @@ static inline _CCCL_DEVICE bool __cuda_atomic_compare_exchange(
     {
       .reg .b128 _d;
       .reg .b128 _v;
-      mov.b128 _d, {%0, %1};
-      mov.b128 _v, {%4, %5};
+      mov.b128 _d, {%3, %4};
+      mov.b128 _v, {%5, %6};
       atom.cas.release.sys.b128 _d,[%2],_d,_v;
       mov.b128 {%0, %1}, _d;
     }
@@ -1813,8 +1813,8 @@ static inline _CCCL_DEVICE bool __cuda_atomic_compare_exchange(
     {
       .reg .b128 _d;
       .reg .b128 _v;
-      mov.b128 _d, {%0, %1};
-      mov.b128 _v, {%4, %5};
+      mov.b128 _d, {%3, %4};
+      mov.b128 _v, {%5, %6};
       atom.cas.acq_rel.cta.b128 _d,[%2],_d,_v;
       mov.b128 {%0, %1}, _d;
     }
@@ -1832,8 +1832,8 @@ static inline _CCCL_DEVICE bool __cuda_atomic_compare_exchange(
     {
       .reg .b128 _d;
       .reg .b128 _v;
-      mov.b128 _d, {%0, %1};
-      mov.b128 _v, {%4, %5};
+      mov.b128 _d, {%3, %4};
+      mov.b128 _v, {%5, %6};
       atom.cas.acq_rel.cluster.b128 _d,[%2],_d,_v;
       mov.b128 {%0, %1}, _d;
     }
@@ -1851,8 +1851,8 @@ static inline _CCCL_DEVICE bool __cuda_atomic_compare_exchange(
     {
       .reg .b128 _d;
       .reg .b128 _v;
-      mov.b128 _d, {%0, %1};
-      mov.b128 _v, {%4, %5};
+      mov.b128 _d, {%3, %4};
+      mov.b128 _v, {%5, %6};
       atom.cas.acq_rel.gpu.b128 _d,[%2],_d,_v;
       mov.b128 {%0, %1}, _d;
     }
@@ -1870,8 +1870,8 @@ static inline _CCCL_DEVICE bool __cuda_atomic_compare_exchange(
     {
       .reg .b128 _d;
       .reg .b128 _v;
-      mov.b128 _d, {%0, %1};
-      mov.b128 _v, {%4, %5};
+      mov.b128 _d, {%3, %4};
+      mov.b128 _v, {%5, %6};
       atom.cas.acq_rel.sys.b128 _d,[%2],_d,_v;
       mov.b128 {%0, %1}, _d;
     }
@@ -1889,8 +1889,8 @@ static inline _CCCL_DEVICE bool __cuda_atomic_compare_exchange(
     {
       .reg .b128 _d;
       .reg .b128 _v;
-      mov.b128 _d, {%0, %1};
-      mov.b128 _v, {%4, %5};
+      mov.b128 _d, {%3, %4};
+      mov.b128 _v, {%5, %6};
       atom.cas.cta.b128 _d,[%2],_d,_v;
       mov.b128 {%0, %1}, _d;
     }
@@ -1908,8 +1908,8 @@ static inline _CCCL_DEVICE bool __cuda_atomic_compare_exchange(
     {
       .reg .b128 _d;
       .reg .b128 _v;
-      mov.b128 _d, {%0, %1};
-      mov.b128 _v, {%4, %5};
+      mov.b128 _d, {%3, %4};
+      mov.b128 _v, {%5, %6};
       atom.cas.cluster.b128 _d,[%2],_d,_v;
       mov.b128 {%0, %1}, _d;
     }
@@ -1927,8 +1927,8 @@ static inline _CCCL_DEVICE bool __cuda_atomic_compare_exchange(
     {
       .reg .b128 _d;
       .reg .b128 _v;
-      mov.b128 _d, {%0, %1};
-      mov.b128 _v, {%4, %5};
+      mov.b128 _d, {%3, %4};
+      mov.b128 _v, {%5, %6};
       atom.cas.gpu.b128 _d,[%2],_d,_v;
       mov.b128 {%0, %1}, _d;
     }
@@ -1946,8 +1946,8 @@ static inline _CCCL_DEVICE bool __cuda_atomic_compare_exchange(
     {
       .reg .b128 _d;
       .reg .b128 _v;
-      mov.b128 _d, {%0, %1};
-      mov.b128 _v, {%4, %5};
+      mov.b128 _d, {%3, %4};
+      mov.b128 _v, {%5, %6};
       atom.cas.sys.b128 _d,[%2],_d,_v;
       mov.b128 {%0, %1}, _d;
     }
--
2.45.2

